File: ueueo-framework/ueueo-eventbus/src/main/java/com/ueueo/eventbus/distributed/InboxProcessor.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using Microsoft.Extensions.Options;
using Volo.Abp.DependencyInjection;
using Volo.Abp.DistributedLocking;
using Volo.Abp.EventBus.Distributed;
using Volo.Abp.Threading;
using Volo.Abp.Timing;
using Volo.Abp.Uow;

namespace Volo.Abp.EventBus.Boxes;

public class InboxProcessor : IInboxProcessor, ITransientDependency
{
    protected IServiceProvider ServiceProvider;//  { get; }
    protected AbpAsyncTimer Timer;//  { get; }
    protected IDistributedEventBus DistributedEventBus;//  { get; }
    protected IAbpDistributedLock DistributedLock;//  { get; }
    protected IUnitOfWorkManager UnitOfWorkManager;//  { get; }
    protected IClock Clock;//  { get; }
    protected IEventInbox Inbox ;// { get; private set; }
    protected InboxConfig InboxConfig ;// { get; private set; }
    protected AbpEventBusBoxesOptions EventBusBoxesOptions;//  { get; }

    protected DateTime? LastCleanTime;// { get; set; }

    protected String DistributedLockName => "AbpInbox_" + InboxConfig.Name;
    public ILogger<InboxProcessor> Logger;// { get; set; }
    protected CancellationTokenSource StoppingTokenSource;//  { get; }
    protected CancellationToken StoppingToken;//  { get; }

    public InboxProcessor(
        IServiceProvider serviceProvider,
        AbpAsyncTimer timer,
        IDistributedEventBus distributedEventBus,
        IAbpDistributedLock distributedLock,
        IUnitOfWorkManager unitOfWorkManager,
        IClock clock,
        IOptions<AbpEventBusBoxesOptions> eventBusBoxesOptions)
    {
        ServiceProvider = serviceProvider;
        Timer = timer;
        DistributedEventBus = distributedEventBus;
        DistributedLock = distributedLock;
        UnitOfWorkManager = unitOfWorkManager;
        Clock = clock;
        EventBusBoxesOptions = eventBusBoxesOptions.Value;
        Timer.Period = Convert.ToInt32(EventBusBoxesOptions.PeriodTimeSpan.TotalMilliseconds);
        Timer.Elapsed += TimerOnElapsed;
        Logger = NullLogger<InboxProcessor>.Instance;
        StoppingTokenSource = new CancellationTokenSource();
        StoppingToken = StoppingTokenSource.Token;
    }

    private void TimerOnElapsed(AbpAsyncTimer arg)
    {
        RunAsync();
    }

    public void StartAsync(InboxConfig inboxConfig, CancellationToken cancellationToken = default)
    {
        InboxConfig = inboxConfig;
        Inbox = (IEventInbox)ServiceProvider.GetRequiredService(inboxConfig.ImplementationType);
        Timer.Start(cancellationToken);
        return Task.CompletedTask;
    }

    public void StopAsync(CancellationToken cancellationToken = default)
    {
        StoppingTokenSource.Cancel();
        Timer.Stop(cancellationToken);
        StoppingTokenSource.Dispose();
        return Task.CompletedTask;
    }

    protected   void RunAsync()
    {
        if (StoppingToken.IsCancellationRequested)
        {
            return;
        }

        using (var handle = DistributedLock.TryAcquireAsync(DistributedLockName, cancellationToken: StoppingToken))
        {
            if (handle != null)
            {
                DeleteOldEventsAsync();

                while (true)
                {
                    var waitingEvents = Inbox.GetWaitingEventsAsync(EventBusBoxesOptions.InboxWaitingEventMaxCount, StoppingToken);
                    if (waitingEvents.Count <= 0)
                    {
                        break;
                    }

                    Logger.LogInformation($"Found {waitingEvents.Count} events in the inbox.");

                    for (var waitingEvent in waitingEvents)
                    {
                        using (var uow = UnitOfWorkManager.Begin(isTransactional: true, requiresNew: true))
                        {
                            DistributedEventBus
                                .AsSupportsEventBoxes()
                                .ProcessFromInboxAsync(waitingEvent, InboxConfig);

                            Inbox.MarkAsProcessedAsync(waitingEvent.Id);

                            uow.CompleteAsync();
                        }

                        Logger.LogInformation($"Processed the incoming event with id = {waitingEvent.Id:N}");
                    }
                }
            }
            else
            {
                Logger.LogDebug("Could not obtain the distributed lock: " + DistributedLockName);
                try
                {
                    Task.Delay(EventBusBoxesOptions.DistributedLockWaitDuration, StoppingToken);
                }
                catch (TaskCanceledException) { }
            }
        }
    }

    protected   void DeleteOldEventsAsync()
    {
        if (LastCleanTime != null && LastCleanTime + EventBusBoxesOptions.CleanOldEventTimeIntervalSpan > Clock.Now)
        {
            return;
        }

        Inbox.DeleteOldEventsAsync();

        LastCleanTime = Clock.Now;
    }
}

File: ueueo-framework/ueueo-eventbus/src/main/java/com/ueueo/eventbus/distributed/InboxProcessor.java
package com.ueueo.eventbus.distributed;

import com.ueueo.distributedlocking.IDistributedLock;
import com.ueueo.distributedlocking.IDistributedLockHandle;
import com.ueueo.threading.CancellationToken;
import com.ueueo.uow.AbpUnitOfWorkOptions;
import com.ueueo.uow.IUnitOfWork;
import com.ueueo.uow.IUnitOfWorkManager;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.BeanFactory;
import org.springframework.beans.factory.config.ConfigurableBeanFactory;
import org.springframework.context.annotation.Scope;
import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
import org.springframework.stereotype.Service;

import java.util.Date;
import java.util.List;

@Service
@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
@Slf4j
public class InboxProcessor implements IInboxProcessor {
    protected BeanFactory beanFactory;
    protected IDistributedEventBus distributedEventBus;
    protected IDistributedLock distributedLock;
    protected IUnitOfWorkManager unitOfWorkManager;
    protected IEventInbox inbox;
    protected InboxConfig inboxConfig;
    protected AbpEventBusBoxesOptions eventBusBoxesOptions;

    protected Date lastCleanTime;

    protected CancellationToken stoppingToken;

    private ThreadPoolTaskScheduler timer;

    protected String getDistributedLockName() {
        return "AbpInbox_" + inboxConfig.getName();
    }

    public InboxProcessor(
            BeanFactory beanFactory,
            ThreadPoolTaskScheduler timer,
            IDistributedEventBus distributedEventBus,
            IDistributedLock distributedLock,
            IUnitOfWorkManager unitOfWorkManager,
            AbpEventBusBoxesOptions eventBusBoxesOptions,
            CancellationToken stoppingToken) {
        this.beanFactory = beanFactory;
        this.timer = timer;
        this.distributedEventBus = distributedEventBus;
        this.distributedLock = distributedLock;
        this.unitOfWorkManager = unitOfWorkManager;
        this.eventBusBoxesOptions = eventBusBoxesOptions;
        this.stoppingToken = stoppingToken;
    }

    @Override
    public void start(InboxConfig inboxConfig) {
        this.inboxConfig = inboxConfig;
        inbox = (IEventInbox) beanFactory.getBean(inboxConfig.getImplementationType());
        timer.scheduleAtFixedRate(this::run, eventBusBoxesOptions.getPeriodTimeSpan().toMillis());
    }

    @Override
    public void stop() {
        stoppingToken.cancel();
        timer.shutdown();
    }

    protected void run() {
        if (stoppingToken.isCancellationRequested()) {
            return;
        }

        IDistributedLockHandle handle = distributedLock.tryAcquire(getDistributedLockName(), eventBusBoxesOptions.getDistributedLockWaitDuration());
        if (handle != null) {
            deleteOldEventsAsync();

            while (true) {
                List<IncomingEventInfo> waitingEvents = inbox.getWaitingEvents(eventBusBoxesOptions.getInboxWaitingEventMaxCount(), stoppingToken);
                if (waitingEvents.size() <= 0) {
                    break;
                }

                log.info("Found {} events in the inbox.", waitingEvents.size());

                for (IncomingEventInfo waitingEvent : waitingEvents) {
                    AbpUnitOfWorkOptions options = new AbpUnitOfWorkOptions();
                    options.setTransactional(true);
                    IUnitOfWork uow = unitOfWorkManager.begin(options, true);

                    distributedEventBus
                            .asSupportsEventBoxes()
                            .processFromInbox(waitingEvent, inboxConfig);

                    inbox.markAsProcessed(waitingEvent.getId());

                    uow.complete();

                    log.info("Processed the incoming event with id = {waitingEvent.Id:N}");
                }
            }
            handle.dispose();
        } else {
            log.debug("Could not obtain the distributed lock: " + getDistributedLockName());
            try {
                Thread.sleep(eventBusBoxesOptions.getDistributedLockWaitDuration().toMillis());
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    protected void deleteOldEventsAsync() {
        if (lastCleanTime != null && lastCleanTime.getTime() + eventBusBoxesOptions.getCleanOldEventTimeIntervalSpan().toMillis() > System.currentTimeMillis()) {
            return;
        }

        inbox.deleteOldEvents();

        lastCleanTime = new Date();
    }
}
